lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1220,6 +1269,220 @@ class CleanupFuncOps : public FuncOpPartialLoweringPattern {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// Simplification patterns
+//===----------------------------------------------------------------------===//
+
+/// Removes operations which have an empty body.
+template <typename TOp>
+struct EliminateEmptyOpPattern : mlir::OpRewritePattern<TOp> {
+  using mlir::OpRewritePattern<TOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.getRegion().empty() || op.getRegion().front().empty()) {
+      rewriter.eraseOp(op);
+      return success();
+    }
+    return failure();
+  }
+};
+
+template <>
+struct EliminateEmptyOpPattern<calyx::IfOp>
+    : mlir::OpRewritePattern<calyx::IfOp> {
+  using mlir::OpRewritePattern<calyx::IfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(calyx::IfOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.thenRegion().empty() || op.thenRegion().front().empty()) {
+      rewriter.eraseOp(op);
+      return success();
+    }
+    return failure();
+  }
+};
+
+/// Removes nested seq operations, i.e.:
```

Comment:
Nit: `e.g.` instead of `.i.e.`, since you're providing an example. Also example can be simplified to one or two lines no?

Reply:
Also, do you foresee this being necessary for `ParOp` as well? For example, the same concept applies for 
```
par { par { ... } } -> par { ... }
```

I think eventually we'll want to move this out as its own pass, as is done in the native compiler [here](https://github.com/cucapra/calyx/blob/dc439d601989c9e72e73288a33904fbc44eac387/calyx/src/passes/collapse_control.rs#L10-L32).

Reply:
Yup. Once the commit series is finished, we should probably take a look at which of these patterns are necessary cleanup (to pass verification) and which are optimization patterns. The latter could (should) be moved to separate passes.